broker: clean shutdown on SIGTERM, add broker.rc2_none attribute

garlick opened the pull request:
This PR preps for a clean Flux shutdown on SIGTERM, which is how systemd will shut down the broker.

First, eliminate the requirement that there is always an initial program.  If the `broker.rc2_none` broker attribute is set, the broker enters rc2 and stays there until...

Second, handle SIGTERM (and other signals) by aborting the current runlevel.  If the runlevel has a subprocess, it is terminated (resulting in the broker exiting with a nozero exit code).  If it doesn't have a subprocess, it simply advances to the next runlevel without error.

The runlevel code was tidied up, some little used runlevel features removed, and minor runlevel bugs fixed.

Combining this PR and #2783 I was finally able to demonstrate saving the KVS content across a systemd stop/start of Flux. 

Edit: description edited to reflect systemd unit file will be updated in another PR.

chu11 commented on src/broker/runlevel.c:
```diff
@@ -196,13 +196,13 @@ static int runlevel_start_subprocess (struct runlevel *r, int level)
         if (flux_subprocess_aux_set (p, "runlevel", r, NULL) < 0)
             goto error;
 
-        monotime (&r->rc[level].start);
```
this removal seems out of place in this commit?

garlick replied:
I think it's _probably_ OK.  The monotime call is moved to after the conditional so the level start time is initialized unconditionally, not just when there is a subprocess.

I included it with the commit that allows rc2 to not have a subprocess because it's noticeable if a garbage time value is printed when rc2 exits.  The fix applies equally to the other run levels, but when there is no subprocess, they are transitioned through instantaneously so time is not interesting, and I think we only do that in test.
```diff
diff --git a/src/broker/runlevel.c b/src/broker/runlevel.c
index c75a8c913..3124e2df6 100644
--- a/src/broker/runlevel.c
+++ b/src/broker/runlevel.c
@@ -196,13 +196,13 @@ static int runlevel_start_subprocess (struct runlevel *r, int level)
         if (flux_subprocess_aux_set (p, "runlevel", r, NULL) < 0)
             goto error;

-        monotime (&r->rc[level].start);
-
         r->rc[level].p = p;
-    } else {
+    }
+    else if (level == 1 || level == 3) {
         if (r->cb)
             r->cb (r, r->level, 0, 0., "Not configured", r->cb_arg);
     }
+    monotime (&r->rc[level].start);
     return 0;

 error:
```
Let me know what you think.  I could split it out easy enough.

garlick replied:
Er, the fix only applies to the rc2 with no subprocess case because when rc1 or rc3 have no subprocess, the callback is made directly with the elapsed time hardwired to zero.

garlick commented:
I was thinking about dropping the systemd unit file changes from this PR and submitting them later in their own PR with the change needed to enable KVS persistence.

garlick commented:
Went ahead and dropped the systemd unit file changes and will submit separately.

garlick commented:
I realized that I was assuming subprocess destructors did not clobber errno in the runlevel code, but on closer inspection, they do.  Rather than wrapping that in the runlevel code, I just fixed it in libsubprocess.  While in there I noticed a couple of unchecked `calloc()` and fixed those.

Fixed a comment typo, rebased on current master, and forced a push.

garlick commented:
I think this PR is ready for review if someone has a moment.  I have another PR that I will submit shortly that takes care of the systemd unit file changes that were taken out of this PR.

If it makes sense for me to submit the libsubprocess cleanup as a separate PR that goes in first LMK.  Might be better for release notes.

grondo commented:
> If it makes sense for me to submit the libsubprocess cleanup as a separate PR that goes in first LMK. Might be better for release notes.

If it isn't too difficult that might be nice.

garlick commented:
> If it isn't too difficult that might be nice.

Done.

I suppose it doesn't matter which one goes in first, since only (uncovered) out-of-memory error paths in this PR would be affected by the other one.

garlick commented:
Thanks!

> I think we already had mentioned that something more subtle might need to be done for ranks > 0 than _exit, but that would definitely be a future PR.

I'll open an issue.  I wasn't entirely sure what the behavior should be.
